tensorboard/data/server/logdir.rs

```diff
@@ -0,0 +1,442 @@
+            let mut result = runs_store
+                .keys()
+                .map(|Run(name)| name.clone())
+                .collect::<Vec<String>>();
+            result.sort();
+            result
+        };
+
+        loader.reload();
+        assert_eq!(get_run_names(), vec!["test", "train"]);
+
+        // Add val, remove train.
+        fs::create_dir(&val_dir)?;
+        File::create(val_dir.join(EVENT_FILE_BASENAME_INFIX))?;
+        fs::remove_file(train_dir.join(EVENT_FILE_BASENAME_INFIX))?;
+        fs::remove_dir(&train_dir)?;
+
+        loader.reload();
+        assert_eq!(get_run_names(), vec!["test", "val"]);
+
+        Ok(())
+    }
+
+    #[cfg(unix)]
+    #[test]
+    fn test_bad_unicode_collision() -> Result<(), Box<dyn std::error::Error>> {
+        use std::os::unix::ffi::OsStrExt;
+
+        // Generate a bad-Unicode collision.
+        let bad1 = Path::new(OsStr::from_bytes(&b"test\x99.run"[..]));
+        let bad2 = Path::new(OsStr::from_bytes(&b"test\xee.run"[..]));
+        let run1 = Run(bad1.to_string_lossy().into_owned());
+        let run2 = Run(bad2.to_string_lossy().into_owned());
+        assert_ne!(bad1, bad2);
+        assert_eq!(run1, run2);
+        let run = run1;
+        drop(run2);
+
+        let logdir = tempfile::tempdir()?;
+        for &dir_basename in &[bad1, bad2] {
+            let dir = logdir.path().join(dir_basename);
+            fs::create_dir(&dir)?;
+            File::create(dir.join(EVENT_FILE_BASENAME_INFIX))?;
+        }
+
+        let commit = Commit::new();
+        let mut loader = LogdirLoader::new(&commit, logdir.path().to_path_buf());
+        loader.reload();
+
+        assert_eq!(loader.runs.keys().collect::<Vec<_>>(), vec![&run]);
+        let run_state = &loader.runs[&run];
+        assert!(bad1 < bad2);
+        assert_eq!(run_state.relpath, bad1);
+        let mut expected_relpaths = HashSet::new();
+        expected_relpaths.insert(bad2.to_path_buf());
+        assert_eq!(run_state.collided_relpaths, expected_relpaths);
+
+        Ok(())
+    }
+
+    #[cfg(unix)]
+    #[test]
+    fn test_symlink() -> Result<(), Box<dyn std::error::Error>> {
+        let logdir = tempfile::tempdir()?;
+        let train_dir = logdir.path().join("train");
+        let test_dir = logdir.path().join("test");
+        fs::create_dir(&train_dir)?;
+        std::os::unix::fs::symlink(&train_dir, &test_dir)?;
+        File::create(train_dir.join(EVENT_FILE_BASENAME_INFIX))?;
+
+        let commit = Commit::new();
+        let mut loader = LogdirLoader::new(&commit, logdir.path().to_path_buf());
+        loader.reload();
+
+        assert_eq!(
+            loader.runs.keys().collect::<HashSet<_>>(),
+            vec![&Run("test".to_string()), &Run("train".to_string())]
+                .into_iter()
+                .collect::<HashSet<_>>(),
+        );
+        Ok(())
+    }
+
+    #[cfg(unix)]
+    #[test]
+    fn test_symlink_loop() -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
Yay :) I was wondering how the cycle is handled and you wrote a test for that!

Reply:
Yep—[`WalkDir::follow_links`] returns an error, which we print and then
continue:

```
$ cargo test --show-output
---- logdir::tests::test_symlink_loop stdout ----
error walking log directory: IO error for operation on /tmp/.tmpAeanMV/dir1: Too many levels of symbolic links (os error 40)
error walking log directory: IO error for operation on /tmp/.tmpAeanMV/dir2: Too many levels of symbolic links (os error 40)

---- logdir::tests::test_bad_unicode_collision stdout ----
merging directories "test\x99.run" and "test\xEE.run", which both normalize to run "test�.run"
```

[`WalkDir::follow_links`]: https://docs.rs/walkdir/2.3.1/walkdir/struct.WalkDir.html#method.follow_links